### PR TITLE
HTTP/2 Thread Context Interface Clarifications

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
It is currently assumed that all usages of the HTTP/2 codec will be from the same event loop context. If the methods are used outside of the assumed thread context then unexpected behavior is observed. This assumption should be more clearly communicated and enforced in key areas.

Modifications:
- The flow controller interfaces have assert statements and updated javadocs indicating the assumptions.

Result:
Interfaces more clearly indicate thread context limitations.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@nmittler - FYI

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java [hunk 1]
```diff
@@ -147,17 +147,43 @@ public ChannelFuture writeData(final ChannelHandlerContext ctx, final int stream
     @Override
     public ChannelFuture writeHeaders(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int padding,
             boolean endStream, ChannelPromise promise) {
-        initCompressor(streamId, headers, endStream);
-        return super.writeHeaders(ctx, streamId, headers, padding, endStream, promise);
+        try {
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java]: @Scottmitch are the changes in this class related to this PR? Are you fixing anything here or just code cleanup?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 2]
```diff
@@ -108,8 +108,15 @@ public void channelHandlerContext(ChannelHandlerContext ctx) {
         this.ctx = ctx;
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method must be called from the same event loop context as the {@link ChannelHandlerContext}
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @normanmaurer @Scottmitch what are your thoughts about using the `@NotThreadSafe` annotation for classes expected to be called within the event loop?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java [hunk 3]
```diff
@@ -133,8 +140,15 @@ public int initialWindowSize(Http2Stream stream) {
         return state(stream).initialWindowSize();
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method must be called from the same event loop context as the {@link ChannelHandlerContext}
```
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @Scottmitch I wonder if we should just document the interfaces to spell out that they should be called in the event loop.  WDYT?

**Comment by nmittler** [comment]
@Scottmitch looks good ... just a couple high-level questions/comments.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java (hunk 1 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java]: Yes there are changes `CompressorHttp2ConnectionEncoder` is broken. The compressor was broken and discovered while do the re-work.  I could break it up into another PR...

Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java]: +1 have two prs sound good

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 3 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @nmittler - The thought was that you could easily have a wrapper or extension which provides the thread safe capability for the user (check if on eventloop, if not execute a runnable). So the interface is not necessarily limited...but instead just this implementation.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 2 above)
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @nmittler - I'm not against it but I'm not crazy about using these annotations either.  For example we don't really use `@ThreadSafe` or `@NotThreadSafe` now, and so if there is no annotation...what should the user assume about this?  I guess the point is we don't necessarily want them to assume anything just because the annotation is not present, but by explicitly using `@NotThreadSafe` they may think if this is not present that the method is thread safe (despite the javadoc on the annotation saying nothing should be assumed).

File: codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java (hunk 1 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java]: What he said :)

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/CompressorHttp2ConnectionEncoder.java]: This PR is now dependent on https://github.com/netty/netty/pull/4000....builds will fail until rebased.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @Scottmitch agreed, we would definitely want some consistency with when `@NotThreadSafe` is used and when it's missing.  Perhaps we could just be consistent within HTTP/2 for now and then tackling the rest of Netty in a follow-up PR? I suspect it would be worthwhile documentation for everyone using Netty.

@normanmaurer any thoughts on using `@NotThreadSafe` in Netty?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 3 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @Scottmitch the HTTP/2 codec is composed of a lot of classes, each of which could be made to be thread-safe. If one of those classes were to be made thread-safe, the HTTP/2 codec would still not be.  It may be fine if the attempt here is to allow external access to one of those classes (e.g. the flow controller). However, if an action on the class causes an interaction with a non-threadsafe class, that would be bad.  Do you have a use-case in mind for making a thread-safe version of one of the classes?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @nmittler - At one point there was interest in providing a "thread safe" wrapper around `Http2FrameWriter`'s `writeXXX` methods.  The idea would be to mimic the thread safety provided by Netty's `channel.write(..)` methods.  I have not gone down this path yet but I'm guessing this would most likely result in a some sort of decorator around `DefaultHttp2ConnectionEncoder` (and maybe useful to provide something similar for `DefaultHttp2FrameWriter` too).

So this PR does not attempt to completely document all the interfaces in the HTTP/2 codec, but instead focuses on the flow controllers.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java (hunk 2 above)
Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: ok makes sense.

Comment (nmittler) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @Scottmitch maybe just document at the class level that all methods should be called in the event loop? Then get rid of the method-by-method docs?

Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2LocalFlowController.java]: @nmittler - sgtm. Will move the docs.

**Comment by Scottmitch** [comment]
@nmittler - PTAL

**Comment by nmittler** [comment]
@Scottmitch LGTM

The only remaining discussion point for me is whether or not to use `@NotThreadSafe` on these classes as well.

@normanmaurer thoughts?

**Comment by normanmaurer** [comment]
@nmittler I would prefer to NOT use `@NotThreadSafe`, at least for now. I think if you really want to use it we should open another issue and discuss it. Would like to hear @trustin on this topic as well.

**Comment by Scottmitch** [comment]
@nmittler - I agree with @normanmaurer. Lets handle the `@NotThreadSafe` as a seperate issue.

**Comment by Scottmitch** [comment]
Cherry-picked.

**Comment by nmittler** [comment]
Sounds good ... handling that separately makes sense.  Although I'm hesitant to open a bug because I don't really care that much :)

**Comment by Scottmitch** [comment]
@nmittler - That is the spirit :)  I'll leave it up to you.

**Comment by nmittler** [comment]
:)
